vyper/context/README.md

```diff
@@ -0,0 +1,325 @@
+example:
+
+```python
+foo: int128 = 0
+bar: uint256 = 0
+
+for i in [1, 2, 3]:
+    foo += i
+    bar += i
+```
+
+1. A definition for `foo` is created with type `int128`.
+2. A definition for `bar` is created with type `uint256`.
+3. A definition for `i` is created. Because no type is explicitly declared, `i` is
+given a `UnionType` of `{int128, uint256}` based on the values of the literal array.
+4. The types `foo` and `i` are compared. The `UnionType` collapses to an `int128`.
+5. The types of `bar` and `i` are compared. Because of the previous comparison `i`
+is known to be type `int128`, so a `TypeMismatch` is raised.
+
+### Definitions
+
+Each definition is represented by an object. Definition classes are created
+dynamically based on the type and value they represent, and the syntax with which
+they are declared.
+
+There are many possible definition types, however all definitions broadly fit
+under two categories.
+
+* **Value** definitions have a value, a type associated with that value.
+* **Callable** definitions have no value, zero or more input types, zero or more
+output types, and a method for call validation.
+
+Because of the dynamic creation of definition classes, it is possible to create a
+definition which is both a value *and* callable.
+
+#### Value Definitions
+
+Value definitions are typically composed of one of two base classes:
+
+* `Literal`: For literals, where an exact value is known
+* `Reference`: For variables or other references where the value may be unknown
+
+This class then inherits any number of base classes which introduce specific
+functionality. The final parent is always `BaseDefinition`.
+
+The `build_value_definition` function handles dynamic creation of value definition
+objects based on a set of AST nodes. The `from_type` classmethod in `Literal` or
+`Reference` is used when a specific base class is requried.
+
+Classes and functions related to value definitions can be found in
+[`definitions/values.py`](definitions/values.py).
+
+#### Callable Definitions
+
+Callable definitions are used to represent anything which is callable, e.g. builtin
+functions, contract functions, interfaces, events, etc.
+
+Callable definitions do not have a type. Instead, they implement the
+`fetch_call_return` method. This method validates a set of call argument AST nodes
+and returns zero or more value definition objects.
+
+Builtin functions may also include an `evaluate` method, used for constant folding
+when the given call args are literals.
+
+#### Annotation Declaration Definitions
+
+Annotation declaration definitions are special definition objects which are declared
+in the annotation field of an `AnnAssign` node. For example, `map`:
+
+```python
+foo: map(int128, address)
+```
+
+See the docstrings in [`definitions/annotation_declaration.py`](definitions/annotation_declaration.py)
+to learn more about creating definitions which are declarable via assignment
+annotations.
+
+## Integration
+
+...
```

Comment:
What will go here?

Reply:
Unclear to me at this time.  in `vyper/ast` it talks about how to import nodes properly.